### DISCUSSION TITLE
Create ASTM/AAMA compliant DXF

### DISCUSSION THREAD

**Discussion by stephenthoma** [mozman/ezdxf#789] [Q&A]
I'm currently running into an issue importing DXFs created by ezdxf into Gerber Accumark (an apparel patterning software). The ASTM standard expects every block in the DXF to be a pattern piece. The layout blocks from ezdxf are causing import to fail.

Is there a way to save DXF blocks outside of a model space layout

<img width="508" alt="Screenshot 2022-11-28 at 10 45 00 PM" src="https://user-images.githubusercontent.com/1490823/204890974-415d94c7-ebce-43dc-85af-86b421c52d14.png">

**Comment by mozman** [comment]
Ezdxf can create every valid DXF structure, it's up to you where to place the expected entities. I need more information to help.

EDIT:

~According to this [post](https://gerbertechnology.vanillacommunity.com/discussion/1305/what-is-the-difference-between-a-standard-dxf-file-and-an-astm-aama-dxf-file) is the ASTM/AAMA standard not compatible to DXF.~

UPDATE: This is just not true, the ASTM/AAMA standard explicit requires DXF compatibility. The problem is that Gerber Technology simply has a very bad DXF parser.

Here is a [solution](https://forums.autodesk.com/t5/autocad-forum/autocad-export-to-dxf-aama-or-dxf-astm/m-p/7027233#M875558) for AutoCAD and can maybe adapted for `ezdxf`.

I found this [PDF](https://dorthehansen.com/wp-content/uploads/2014/10/ePattern-ASTM-Standard.pdf) online: This document presents readers with an overview of the technical aspects related to the contents and structure of digital patterns according to ASTM standard D6673 / D6963.

As many official standards the ASTM D6673-10 standard is not free for download. The way I see it, organizations that don't release their standards for free don't want support from the open source community.

**Reply by michael-p-sachen** [reply]
On a tangent for our specific case,
We are having an issue in that ezdxf stores all blocks with reference to a block layout. When we use these outputs in a lot of our software it's not able to parse this dxf structure.

On the other hand, when we export from these programs to ASTM/AAMA complaint dxfs, the blocks live under a blocks section (I've included a dxf to demonstrate the structure) without reference to a layout

I've tried string parsing the geometry out of the ezdxf files to match this expectation, but have not found it elegant nor have had luck with it...
really we are looking to capture just the block geometry only

[square.dxf.txt](https://github.com/mozman/ezdxf/files/10133215/square.dxf.txt)

**Reply by mozman** [reply]
I'm not sure if the problems are caused by the ASTM DXF format itself or by some crappy implementation by Gerber Technology. I have already a "special" DXF file from Gerber Technology with invalid content for testing. The ASTM-DXF format does not prohibit layout blocks and could easily be ignored, as they start with an otherwise invalid character "*".

The sample file also does not follow the format rules for SST and PST:

ASTM defines the descriptive text that is valid to all pieces in a pattern as “Style System Text”. Style 
System Text (SST) is multiline text placed on layer 1. The SST is mandatory. It is not to be confused 
with piece system text or general annotation. The text follows a **precise syntax which is case sensitive 
in a mix of upper and lower case characters** in ASCII-7 (standard keyboard). Style System Text can be 
viewed on layer 1 of sample ASTM files. 

EDIT: 
Cite from the official D6673-10 standard: Identiﬁers can be formed from upper or lower case characters but must appear as speciﬁed in this standard.

But is all uppercase in the sample file:

![image](https://user-images.githubusercontent.com/4230966/205235661-fd3b2871-0f8f-4753-82e7-67d813d5574f.png)

EDIT: the sample file can be open by BricsCAD and ezdxf, but not by Autodesk Trueview - ~invalid DXF file!~

Some additional citations from the D6673-10 standard:

  1.2  The ﬁle format for the pattern data exchange ﬁle deﬁned
  by this standard (Practice D6673) complies with the Drawing
  Interchange File (DXF) format. Autodesk, Inc. developed the
  DXF format for transferring data between their AutoCAD(r)
  product and other software applications. This standard docu-
  ments the manner in which pattern data should be represented
  within  the  DXF  format.  Users  of  this  standard  should  have
  Autodesk, Inc.’s documentation on Drawing Interchange Files
  found in the AutoCAD Reference Manual, in order to assure
  compatibility to  all DXF format speciﬁcations. The AutoCAD
  Version 13 DXF speciﬁcation is to be used. The ﬁle format for
  the grade rule table exchange ﬁle is an ASCII text ﬁle.

~The sample file  square.dxf does not correspond to the DXF standard!~

This is more an issue of Autodesk Trueview, see posts below!

**Comment by mozman** [comment]
I am not sure what you want.  The example file contains a block definition and a block reference to that block:

![image](https://user-images.githubusercontent.com/4230966/205100526-cec206d7-7cb7-4ef9-b8de-2206295a345c.png)

- Is this the desired output or is this the output that doesn't work?
- What do you want to extract?

**Reply by michael-p-sachen** [reply]
Hi, the file that I included is the desired format. 
The issue we are having is that we can not output from ezdxf in a way that is readable with our software.

It complains that Modelspace/Paperspace entities are not valid ASTM/AAMA blocks
And we are trying to find a way to export blocks from ezdxf without modelspace/paperspace references

**Comment by mozman** [comment]
Okay, I understand what you need, but this is not possible by `ezdxf`, the modelspace and the active paperspace layout are so fundamentally required for the DXF format, that the export process do not work if they do not exist. It's possible to delete all layouts but the export process raises an error:

```Python
import ezdxf

doc = ezdxf.readfile("square.dxf")
# this is a hack, don't try this at home! 😄
# delete the modelspace and all paperspace layouts
doc.layouts._layouts.clear()  
doc.saveas("raises_an_exeption.dxf")
```

**Reply by mozman** [reply]
BTW: ~this is the worst case of a format misuse if have ever seen.~ That's not meant to be funny.

Gerber Technology simply has a very bad DXF parser.

**Comment by mozman** [comment]
I hacked an "epattern" add-on in the `astm_d6673` branch, which can load and export this format: 

https://github.com/mozman/ezdxf/blob/astm_d6673/src/ezdxf/addons/epattern.py

```Python
from ezdxf.addons import epattern

doc = epattern.readfile("square.dxf")
# doc = epattern.new() to create a new document
doc.saveas("test.dxf")


```

Let me hear if this works for you, also if "epattern" is a suitable name. ASTM is a standard organization, what does AAMA mean?

**Reply by michael-p-sachen** [reply]
Thanks! Trying it now... its the American Apparel Manufacturing Association

**Reply by mozman** [reply]
I refactored the add-on, so no special `Drawing` class is need:

```Python
import ezdxf
from ezdxf.addons import gerber_D6673

# use the regular loading functions of ezdxf or the recover module to load the "Gerber" files
doc = ezdxf.readfile("square.dxf")

# export DXF R12 documents as special "Gerber" files that claim to be ASTM-D6673-10 compliant:
gerber_D6673.export_file(doc, "test.dxf")
```

**Reply by mozman** [reply]
I added some annotations to the sample file provided. Reading the D6673-10 standard, it becomes clear that this add-on is not required to create DXF files that complie to the D6673-10 standard, its sole purpose is to generate ~invalid~ special DXF files which can be read by a poorly written software.

~I'm not ready to merge this add-on into the master branch.~

**Comment by mozman** [comment]
SUMMARY
------------

~Gerber Technology software does not conform to the ASTM D6673-10 standard, which requires a valid DXF format. The sample file provided cannot be opened in an Autodesk product, which clearly indicates that the file is not a DXF file. There is no need to change anything in `ezdxf` and I know it won't work, but send a bug report to Gerber Technology to fix their software - this should solve this problem.~

Gerber Technology applications just have an low-quality DXF parser and this add-on can help to feed them the expected data.

**Reply by mozman** [reply]
After sleeping one night over this problem, I decided to help Gerber Technology application users. The add-on is merged into the master branch under the name "gerber_D6673.py".

~I'm still upset with Gerber Technology, but that's not the fault of the users who want or need to use their software.~

**Reply by mozman** [reply]
A lot of noise about nothing!

The "Gerber" files are valid DXF R12 files, Autodesk TrueView complains only about an invalid BLOCK name "Shape 0_M", which in my opinion is valid, maybe spaces were not allowed in the original R12 version, but this is just a minor issue and is more a problem of the picky Autodesk DXF parser, which is otherwise very forgiving for the DXF R12 Standard.

**Comment by michael-p-sachen** [comment]
Hi! Thank you for all of this. We are waiting to hear back from our factory that uses Gerber to ensure that everything is running well now. 

I will update you here
